File: Sources/Funnybot/AskGpt/AskGpt/AskGpt.swift
import Foundation
import Schwifty

open class AskGpt {
    private let chatApi: ChatApi
    
    public init(apiKey: String, model: AskGptModel) {
        Logger.debug("AskGpt", "Init with model:", model.name)
        self.chatApi = model.buildChatApi(apiKey: apiKey)
    }
    
    public func ask(prompt: String, question: String) async throws -> String {
        let history = [
            ChatMessage(role: .system, name: "GPT", content: prompt),
            ChatMessage(role: .user, name: "User", content: question)
        ]
        Logger.debug("AskGpt", "System Prompt: \(prompt)")
        Logger.debug("AskGpt", "Question: \(question)")
        Logger.debug("AskGpt", "Waiting for response...")
        
        let result = await chatApi.next(from: history)
        switch result {
        case .success(let success):
            Logger.debug("AskGpt", "Answer: \(success.content)")
            return success.content
        case .failure(let error):
            Logger.debug("AskGpt", "No answer! \(error)")
            throw error
        }
    }
    
    public func response(for history: [ChatMessage], prompt: String) async throws -> String {
        Logger.debug("AskGpt", "Prompt: \(prompt)")
        Logger.debug("AskGpt", "Waiting for response...")
        
        let fullHistory = history + [ChatMessage(role: .system, name: "System", content: prompt)]
        
        let result = await chatApi.next(from: fullHistory)
        switch result {
        case .success(let success):
            Logger.debug("AskGpt", "Answer: \(success.content)")
            return success.content
        case .failure(let error):
            Logger.debug("AskGpt", "No answer! \(error)")
            throw error
        }
    }
}

File: Sources/Scriptz/askgpt.py
from random import random
from time import sleep

import openai

GPT35_16 = "gpt-3.5-turbo-16k"
GPT35 = "gpt-3.5-turbo"
GPT4 = "gpt-4"
openai.api_key = "..."


def list_models():
    models = openai.Model.list()
    f = open("models.json", "w")
    f.write(str(models))
    f.close()


def response(history, model=GPT4):
    completion = openai.ChatCompletion.create(
        model=model, messages=history, max_tokens=2000
    )
    response = completion.choices[0].message.content
    return response


def ask(initial_prompt, prompts, model=GPT35):
    history = [{"role": "system", "content": initial_prompt}]
    responses = []

    for i, prompt in enumerate(prompts):
        if prompt.startswith("$$$nochange:"):
            responses.append(prompt)
            continue
        if i % 10 == 0:
            sleep(random())

        new_message = {"role": "user", "content": prompt}
        result = response(history + [new_message], model)
        responses.append(result)

    return responses


if __name__ == "__main__":
    list_models()
